File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/client/BridgeProofSubmitRequest.java
package org.hyperledger.iroha.android.client;

import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/** Exact first-release request payload for {@code POST /v1/bridge/proofs/submit}. */
public final class BridgeProofSubmitRequest {
  private final String authority;
  private final String publicKeyHex;
  private final String signatureB64;
  private final String messageBundleB64;
  private final String networkIdHex;
  private final String verifierAddressHex;
  private final String bridgeAddressHex;
  private final String verifierCodeHashHex;
  private final String verifierKeyHashHex;
  private final String tronVerifierAddress;
  private final String proofBytesHex;
  private final Long creationTimeMs;

  private BridgeProofSubmitRequest(final Builder builder) {
    authority = SccpSubmitEncoding.requireCanonicalNonBlank(builder.authority, "authority");
    publicKeyHex = SccpSubmitEncoding.normalizeOptionalPublicKeyHex(builder.publicKeyHex);
    signatureB64 =
        SccpSubmitEncoding.normalizeOptionalExactBase64(builder.signatureB64, "signatureB64");
    messageBundleB64 = builder.messageBundleB64;
    SccpSubmitEncoding.validateCanonicalNoritoBase64(
        messageBundleB64, "messageBundleB64", SccpSubmitEncoding.MAX_ARTIFACT_BYTES);
    networkIdHex = SccpSubmitEncoding.normalizeOptionalHex(builder.networkIdHex, 32, "networkIdHex");
    verifierAddressHex =
        SccpSubmitEncoding.normalizeOptionalHex(
            builder.verifierAddressHex, 20, "verifierAddressHex");
    bridgeAddressHex =
        SccpSubmitEncoding.normalizeOptionalHex(builder.bridgeAddressHex, 20, "bridgeAddressHex");
    verifierCodeHashHex =
        SccpSubmitEncoding.normalizeOptionalHex(
            builder.verifierCodeHashHex, 32, "verifierCodeHashHex");
    verifierKeyHashHex =
        SccpSubmitEncoding.normalizeOptionalHex(
            builder.verifierKeyHashHex, 32, "verifierKeyHashHex");
    tronVerifierAddress = SccpSubmitEncoding.normalizeOptional(builder.tronVerifierAddress);
    proofBytesHex = SccpSubmitEncoding.normalizeOptional(builder.proofBytesHex);
    creationTimeMs = SccpSubmitEncoding.normalizeOptionalCreationTimeMs(builder.creationTimeMs);

    if ((publicKeyHex == null) != (signatureB64 == null)) {
      throw new IllegalArgumentException("publicKeyHex and signatureB64 must be supplied together");
    }

    final boolean destinationPresent = hasDestinationMaterial();
    if ((proofBytesHex == null) != !destinationPresent) {
      throw new IllegalArgumentException(
          "proofBytesHex and complete destination material must be supplied together");
    }
    if (destinationPresent) {
      requireCompleteDestinationTuple();
    }
  }

  public String authority() {
    return authority;
  }

  public String messageBundleB64() {
    return messageBundleB64;
  }

  public Map<String, Object> toJsonMap() {
    final Map<String, Object> json = new LinkedHashMap<>();
    json.put("authority", authority);
    putIfPresent(json, "public_key_hex", publicKeyHex);
    putIfPresent(json, "signature_b64", signatureB64);
    putIfPresent(json, "message_bundle_b64", messageBundleB64);
    putIfPresent(json, "network_id_hex", networkIdHex);
    putIfPresent(json, "verifier_address_hex", verifierAddressHex);
    putIfPresent(json, "bridge_address_hex", bridgeAddressHex);
    putIfPresent(json, "verifier_code_hash_hex", verifierCodeHashHex);
    putIfPresent(json, "verifier_key_hash_hex", verifierKeyHashHex);
    putIfPresent(json, "tron_verifier_address", tronVerifierAddress);
    putIfPresent(json, "proof_bytes_hex", proofBytesHex);
    putIfPresent(json, "creation_time_ms", creationTimeMs);
    return Collections.unmodifiableMap(json);
  }

  public byte[] toJsonBytes() {
    return JsonEncoder.encode(toJsonMap()).getBytes(StandardCharsets.UTF_8);
  }

  public static Builder builder() {
    return new Builder();
  }

  private boolean hasDestinationMaterial() {
    return networkIdHex != null
        || verifierAddressHex != null
        || bridgeAddressHex != null
        || verifierCodeHashHex != null
        || verifierKeyHashHex != null
        || tronVerifierAddress != null;
  }

  private void requireCompleteDestinationTuple() {
    final boolean evm = verifierAddressHex != null || bridgeAddressHex != null;
    final boolean tron = tronVerifierAddress != null;
    if (evm == tron) {
      throw new IllegalArgumentException(
          "destination material must select exactly one EVM or TRON family");
    }
    if (networkIdHex == null || verifierCodeHashHex == null || verifierKeyHashHex == null) {
      throw new IllegalArgumentException("complete SCCP destination material is required");
    }
    if (evm && (verifierAddressHex == null || bridgeAddressHex == null)) {
      throw new IllegalArgumentException("complete EVM SCCP destination material is required");
    }
    if (tron && tronVerifierAddress.isEmpty()) {
      throw new IllegalArgumentException("complete TRON SCCP destination material is required");
    }
  }

  private static void putIfPresent(
      final Map<String, Object> target, final String key, final Object value) {
    if (value != null) {
      target.put(key, value);
    }
  }

  /** Builder for the exact request shape. */
  public static final class Builder {
    private String authority;
    private String publicKeyHex;
    private String signatureB64;
    private String messageBundleB64;
    private String networkIdHex;
    private String verifierAddressHex;
    private String bridgeAddressHex;
    private String verifierCodeHashHex;
    private String verifierKeyHashHex;
    private String tronVerifierAddress;
    private String proofBytesHex;
    private Long creationTimeMs;

    public Builder authority(final String value) {
      authority = value;
      return this;
    }

    public Builder publicKeyHex(final String value) {
      publicKeyHex = value;
      return this;
    }

    public Builder signatureB64(final String value) {
      signatureB64 = value;
      return this;
    }

    public Builder messageBundleB64(final String value) {
      messageBundleB64 = value;
      return this;
    }

    public Builder networkIdHex(final String value) {
      networkIdHex = value;
      return this;
    }

    public Builder verifierAddressHex(final String value) {
      verifierAddressHex = value;
      return this;
    }

    public Builder bridgeAddressHex(final String value) {
      bridgeAddressHex = value;
      return this;
    }

    public Builder verifierCodeHashHex(final String value) {
      verifierCodeHashHex = value;
      return this;
    }

    public Builder verifierKeyHashHex(final String value) {
      verifierKeyHashHex = value;
      return this;
    }

    public Builder tronVerifierAddress(final String value) {
      tronVerifierAddress = value;
      return this;
    }

    public Builder proofBytesHex(final String value) {
      proofBytesHex = value;
      return this;
    }

    public Builder creationTimeMs(final Long value) {
      creationTimeMs = value;
      return this;
    }

    public BridgeProofSubmitRequest build() {
      return new BridgeProofSubmitRequest(this);
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/client/BridgeProofSubmitRequest.kt
package org.hyperledger.iroha.sdk.client

import java.util.Base64
import org.hyperledger.iroha.sdk.norito.NoritoHeader

/** Request payload for `POST /v1/bridge/proofs/submit`. */
class BridgeProofSubmitRequest(
    authority: String,
    messageBundleB64: String,
    publicKeyHex: String? = null,
    signatureB64: String? = null,
    networkIdHex: String? = null,
    verifierAddressHex: String? = null,
    bridgeAddressHex: String? = null,
    verifierCodeHashHex: String? = null,
    verifierKeyHashHex: String? = null,
    tronVerifierAddress: String? = null,
    proofBytesHex: String? = null,
    creationTimeMs: Long? = null,
) {
    val authority: String = requireNonBlank(authority, "authority")
    val publicKeyHex: String? = normalizeOptionalPublicKeyHex(publicKeyHex)
    val signatureB64: String? = normalizeOptionalExactBase64(signatureB64, "signatureB64")
    val messageBundleB64: String = messageBundleB64.also {
        validateCanonicalSccpNoritoBase64(it, "messageBundleB64", SCCP_MAX_MESSAGE_BUNDLE_BYTES)
    }
    val networkIdHex: String? = normalizeOptionalHex(networkIdHex, 32, "networkIdHex")
    val verifierAddressHex: String? = normalizeOptionalHex(verifierAddressHex, 20, "verifierAddressHex")
    val bridgeAddressHex: String? = normalizeOptionalHex(bridgeAddressHex, 20, "bridgeAddressHex")
    val verifierCodeHashHex: String? = normalizeOptionalHex(verifierCodeHashHex, 32, "verifierCodeHashHex")
    val verifierKeyHashHex: String? = normalizeOptionalHex(verifierKeyHashHex, 32, "verifierKeyHashHex")
    val tronVerifierAddress: String? = normalizeOptional(tronVerifierAddress)
    val proofBytesHex: String? = normalizeOptional(proofBytesHex)
    val creationTimeMs: Long? = creationTimeMs?.also {
        require(it > 0) { "creationTimeMs must be positive" }
    }

    init {
        require((this.publicKeyHex == null) == (this.signatureB64 == null)) {
            "publicKeyHex and signatureB64 must be supplied together"
        }
        val destinationPresent = hasDestinationMaterial()
        require((this.proofBytesHex == null) == !destinationPresent) {
            "proofBytesHex and complete destination material must be supplied together"
        }
        if (destinationPresent) requireCompleteDestinationTuple()
    }

    fun toJsonMap(): Map<String, Any?> = buildMap {
        put("authority", authority)
        publicKeyHex?.let { put("public_key_hex", it) }
        signatureB64?.let { put("signature_b64", it) }
        put("message_bundle_b64", messageBundleB64)
        networkIdHex?.let { put("network_id_hex", it) }
        verifierAddressHex?.let { put("verifier_address_hex", it) }
        bridgeAddressHex?.let { put("bridge_address_hex", it) }
        verifierCodeHashHex?.let { put("verifier_code_hash_hex", it) }
        verifierKeyHashHex?.let { put("verifier_key_hash_hex", it) }
        tronVerifierAddress?.let { put("tron_verifier_address", it) }
        proofBytesHex?.let { put("proof_bytes_hex", it) }
        creationTimeMs?.let { put("creation_time_ms", it) }
    }

    fun toJsonBytes(): ByteArray = JsonEncoder.encode(toJsonMap()).toByteArray(Charsets.UTF_8)

    private fun hasDestinationMaterial(): Boolean =
        networkIdHex != null || verifierAddressHex != null || bridgeAddressHex != null ||
            verifierCodeHashHex != null || verifierKeyHashHex != null || tronVerifierAddress != null

    private fun requireCompleteDestinationTuple() {
        val evm = verifierAddressHex != null || bridgeAddressHex != null
        val tron = tronVerifierAddress != null
        require(evm != tron) { "destination material must select exactly one EVM or TRON family" }
        val commonPresent = networkIdHex != null && verifierCodeHashHex != null && verifierKeyHashHex != null
        require(commonPresent) { "complete SCCP destination material is required" }
        if (evm) {
            require(verifierAddressHex != null && bridgeAddressHex != null) {
                "complete EVM SCCP destination material is required"
            }
        } else {
            require(!tronVerifierAddress.isNullOrBlank()) { "complete TRON SCCP destination material is required" }
        }
    }
}

/** Native-proof-only settlement request for `POST /v1/bridge/messages`. */
class BridgeMessageSubmitRequest(
    authority: String,
    nativeProofB64: String,
    publicKeyHex: String? = null,
    signatureB64: String? = null,
    creationTimeMs: Long? = null,
) {
    val authority: String = requireNonBlank(authority, "authority")
    val publicKeyHex: String? = normalizeOptionalPublicKeyHex(publicKeyHex)
    val signatureB64: String? = normalizeOptionalExactBase64(signatureB64, "signatureB64")
    val nativeProofB64: String = nativeProofB64.also {
        validateCanonicalSccpNoritoBase64(it, "nativeProofB64", SCCP_MAX_NATIVE_PROOF_BYTES)
    }
    val creationTimeMs: Long? = creationTimeMs?.also {
        require(it > 0) { "creationTimeMs must be positive" }
    }

    init {
        require((this.publicKeyHex == null) == (this.signatureB64 == null)) {
            "publicKeyHex and signatureB64 must be supplied together"
        }
    }

    fun toJsonMap(): Map<String, Any?> = buildMap {
        put("authority", authority)
        publicKeyHex?.let { put("public_key_hex", it) }
        signatureB64?.let { put("signature_b64", it) }
        put("native_proof_b64", nativeProofB64)
        creationTimeMs?.let { put("creation_time_ms", it) }
    }

    fun toJsonBytes(): ByteArray = JsonEncoder.encode(toJsonMap()).toByteArray(Charsets.UTF_8)
}

private const val SCCP_MAX_MESSAGE_BUNDLE_BYTES = 16 * 1024 * 1024
private const val SCCP_MAX_NATIVE_PROOF_BYTES = 16 * 1024 * 1024

internal fun validateCanonicalSccpNoritoBase64(value: String, field: String, maximum: Int): ByteArray {
    require(value.isNotEmpty() && value == value.trim()) { "$field must be canonical padded base64" }
    val decoded = try {
        Base64.getDecoder().decode(value)
    } catch (ex: IllegalArgumentException) {
        throw IllegalArgumentException("$field must be valid base64", ex)
    }
    require(decoded.isNotEmpty() && decoded.size <= maximum) { "$field exceeds its canonical size bound" }
    require(Base64.getEncoder().encodeToString(decoded) == value) { "$field must be canonical padded base64" }
    val result = try {
        NoritoHeader.decode(decoded, null)
    } catch (ex: IllegalArgumentException) {
        throw IllegalArgumentException("$field must contain a canonical Norito envelope", ex)
    }
    val header = result.header
    require(header.compression == NoritoHeader.COMPRESSION_NONE) { "$field must use uncompressed canonical Norito" }
    val headerPadding = decoded.size - NoritoHeader.HEADER_LENGTH - header.payloadLength
    require(headerPadding == 0 || headerPadding == 8) {
        "$field must use canonical Norito header alignment padding"
    }
    require(header.schemaHash.any { it.toInt() != 0 }) { "$field must advertise a nonzero Norito schema" }
    require(header.encode().contentEquals(decoded.copyOfRange(0, NoritoHeader.HEADER_LENGTH))) {
        "$field contains a non-canonical Norito header"
    }
    header.validateChecksum(result.payload)
    return decoded
}

private fun requireNonBlank(value: String, field: String): String {
    require(value.isNotBlank() && value == value.trim()) { "$field is required and must be canonical" }
    return value
}

private fun normalizeOptional(value: String?): String? {
    if (value == null) return null
    require(value == value.trim()) { "optional string fields must not contain surrounding whitespace" }
    return value.ifEmpty { null }
}

private fun normalizeOptionalHex(value: String?, bytes: Int, field: String): String? {
    val normalized = normalizeOptional(value) ?: return null
    require(normalized.startsWith("0x") && normalized.length == 2 + bytes * 2) {
        "$field must be canonical lowercase 0x-prefixed $bytes-byte hex"
    }
    require(normalized.substring(2).all { it in '0'..'9' || it in 'a'..'f' }) {
        "$field must be canonical lowercase 0x-prefixed $bytes-byte hex"
    }
    require(normalized.substring(2).any { it != '0' }) { "$field must be nonzero" }
    return normalized
}

private fun normalizeOptionalPublicKeyHex(value: String?): String? {
    val normalized = normalizeOptional(value) ?: return null
    require(Regex("[0-9a-f]{64}").matches(normalized) && normalized.any { it != '0' }) {
        "publicKeyHex must be exactly 32 nonzero lowercase hexadecimal bytes"
    }
    return normalized
}

private fun normalizeOptionalExactBase64(value: String?, field: String): String? {
    if (value == null) return null
    require(value.isNotEmpty() && value == value.trim()) { "$field must be exact standard-base64" }
    val decoded = try {
        Base64.getDecoder().decode(value)
    } catch (ex: IllegalArgumentException) {
        throw IllegalArgumentException("$field must be valid base64", ex)
    }
    require(decoded.isNotEmpty()) { "$field must not decode to empty bytes" }
    if (field == "signatureB64") require(decoded.size == 64) { "$field must contain a 64-byte Ed25519 signature" }
    require(Base64.getEncoder().encodeToString(decoded) == value) { "$field must be exact standard-base64" }
    return value
}
